### ISSUE TITLE
Partial update events

### ISSUE THREAD

**Issue by RationalCoding** [automerge/hypermerge#6]
The typical usecase for a CRDT library like this would be to sync a "vanilla" plaintext document model over the network. This library looks great (the network being handled by DAT is brilliant), but it's missing observers for fine-grained changes.

`hm.on('document:updated', docId, doc)` partially meets this, but there's no indication of what has changed in the document, so I need to store a previous version of the document and diff them every change (which is infeasible).

What's needed is an update event that gives the range removed, old text, and new text. Batching these events would be a plus.

**Comment by jeffpeterson** [comment]
Aha, interesting.

So, the way we can get a (useful) diff of what has changed is via `Automerge.diff(oldDoc, newDoc)`. See the [docs](https://github.com/automerge/automerge#examining-document-history) for the output format.

Keep in mind that this is `O(n)` with respect to the number of _additional_ changes in the new document. It doesn't do a structural diff of the entire document, so it should be pretty performant.

I could add the previous version of the doc to the event if it makes this simpler:

```js
hm.on('document:updated', (docId, doc, pDoc) => {
  const diff = Automerge.diff(pDoc, doc)
  // Do something with diff...
})
```

**Comment by jeffpeterson** [comment]
Yep. Under the hood (`doc._state`), Automerge maintains a list of all changes that have been applied to the doc. `Automerge.diff` simply fetches any changes missing from the old doc and returns them in a developer-friendly format.

Keep in mind, if you want diffs within strings, they'll need to be stored as [`Automerge.Text`](https://github.com/automerge/automerge#text-editing-support).

I'll go ahead and publish a new version with that change, I think it'd be generally useful for other folks as well.

**Comment by jeffpeterson** [comment]
Multihack looks cool! I should point out, though (which evidently the Readme does not) that hypermerge only works in node :/ Dat's [`discovery-swarm`](https://github.com/mafintosh/discovery-swarm) requires opening TCP sockets, which can't be done in-browser.

Automerge itself works fine in the browser, and given a socket connection to a server or other clients, can be wired up fairly easily with [`Automerge.Connection`](https://github.com/automerge/automerge#sending-and-receiving-changes). But, it's definitely more work to integrate than hypermerge.

But, FWIW, the new version is published: `hypermerge@0.4.0`.
